modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-renderer/src/main/resources/META-INF/resources/js/hooks/useStorage.es.js

```diff
@@ -0,0 +1,49 @@
+/**
+ * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License as published by the Free
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+import {useEffect} from 'react';
+
+const defaultDataStorage = {
+	components: new Map(),
+	refs: 0,
+};
+
+const DDMFormRenderer = Symbol('ddm.form.internal');
+
+/**
+ * Simple implementation to create a global context that can be shared
+ * independently of the React tree, Form Renderer is not an application
+ * and can be reused more than once on the same page, so we need to share
+ * requests and fields, they are loaded on demand, so as not to make
+ * unnecessary requests. Use the `useStorage` hook as a way to cache data.
+ */
+export const useStorage = () => {
+	if (!window[DDMFormRenderer]) {
+		window[DDMFormRenderer] = defaultDataStorage;
+	}
+
+	useEffect(() => {
+		window[DDMFormRenderer].refs += 1;
+
+		return () => {
+			window[DDMFormRenderer].refs -= 1;
+
+			if (window[DDMFormRenderer].refs === 0) {
+				delete window[DDMFormRenderer];
```

Comment:
The ref-counting thing seems like complexity that may not be justified, because at this point we're talking about a single, empty `Map` instance.

Reply:
This is also used for the `useResource` cache, I didn’t want to keep the `DDMFormRenderer` property instantiated in the window when browsing out of DDM. So this is more of a validator for removing information whenever no one else is using it.

Maybe not using the `window` anymore we don't need that anymore. But I'm wondering if this data will persist after browsing back to DDM in case I use Senna. Liferay.loader will return the same instance of the source code, as the AMD Loader caches. Does it come back again?

Reply:
I would expect the previously loaded module to remain in memory and not be reloaded/re-evaluated, so any internal data structures declared at the module level will remain and be re-used.

Reply:
Makes sense. The concern is that there is some leakage of data or information during navigation and it is difficult for the GC to collect. So I ended up removing it just to make sure that during browsing it didn't occupy much memory.